jcstress-core/src/main/java/org/openjdk/jcstress/infra/grading/FailFastKiller.java

```diff
@@ -0,0 +1,172 @@
+        this.relative = originalValue.endsWith("%");
+        this.superRelative = originalValue.endsWith("%%");
+        this.isFailFastAllVariants = opts.isFailFastAllVariants();
+        this.variants = Collections.unmodifiableList(finalVariants);
+        this.tests = groupVariants(finalVariants);
+        if (isFailFastAllVariants) {
+            output.println("    all variants, " + originalValue);
+            usedTotal = variants.size();
+        } else {
+            output.println("    whole tests, " + originalValue);
+            usedTotal = tests.size();
+        }
+        if (superRelative) {
+            relativeThreshold = userValue;
+            absoluteThreshold = -1;
+            output.println("    The suite will terminate once failure rate reaches " + getRelativeThresholdNice()
+                    + "% of *currently* finished number of tests/variants");
+        } else {
+            if (relative) {
+                relativeThreshold = userValue;
+                if (isFailFastAllVariants) {
+                    absoluteThreshold = (relativeThreshold * (double) variants.size()) / 100d;
+                } else {
+                    absoluteThreshold = (relativeThreshold * (double) tests.size()) / 100d;
+                }
+            } else {
+                absoluteThreshold = (long) userValue;
+                if (isFailFastAllVariants) {
+                    relativeThreshold = (absoluteThreshold * 100d) / (double) variants.size();
+                } else {
+                    relativeThreshold = (absoluteThreshold * 100d) / (double) tests.size();
+                }
+            }
+            output.println("    The suite will terminate once failure rate reaches " + getRelativeThresholdNice() + "% ("
+                    + getAbsoluteThresholdNice() + ") of total tests/variants (" + usedTotal + ")");
+        }
+        output.println();
+    }
+
+    private static Map<String, Integer> groupVariants(List<TestConfig> finalVariants) {
+        Map<String, Integer> tests = new HashMap<>();
+        for (TestConfig testVariant : finalVariants) {
+            int counter = tests.getOrDefault(testVariant.name, 0);
+            counter++;
+            tests.put(testVariant.name, counter);
+        }
+        return tests;
+    }
+
+    @Override
+    public synchronized void add(TestResult r) {
+        addResult(r);
+    }
+
+    private void addResult(TestResult r) {
+        int groupCounter = tests.get(r.getName());
+        //we are counting each group down, so we can assure that the group is finished
+        tests.put(r.getName(), groupCounter-1);
+        long wasFailed = hardErrors + failed;
+        countResult(r);
+        long isFailed = hardErrors + failed;
+        if (isFailed > wasFailed) {
+            failures.add(r.getName());
+        }
+        verifyState(r);
+    }
+
+    private void verifyState(TestResult r) {
+        long totalFailed = failed + hardErrors;
```

Comment:
Maybe the `softErrors`should be counted in? Maybe configurable by some property? wdyt?

Reply:
done